# Lobito Corridor Spatial Analysis
----

Benny Istanto, bistanto@worldbank.org, GOST/DEC Data Group

Objective. Produce fast, reproducible **access & opportunity** analytics using only aligned rasters: 
1. Isochrone coverage KPIs (population, cropland, electrification), 
2. A transparent **priority surface** for smallholder aggregation/feeder upgrades, 
3. A **flood-aware road risk** screening, and 
4. **Site cards** for the project locations.
              
**Expected outputs.** AOI-prefixed GeoTIFFs in `outputs/rasters/`, CSV tables in `outputs/tables/`, and quick PNG maps in `outputs/figs/`.

## Run once after (re)starting the kernel


<div class="alert alert-block alert-success" align="left">
  <b>Why this cell?</b><br/>
  It initializes the workspace for the current AOI—sets <code>PROJECT_ROOT</code>/<code>AOI</code>, adds <code>src/</code> to <code>sys.path</code>, clears the import cache, and loads <code>config.py</code> so <code>PATHS</code>/<code>PARAMS</code> are available to all steps.
  <br/><br/>
  <b>When to run it</b>
  <ul>
    <li>After a kernel restart (once).</li>
    <li>After changing <code>AOI_VALUE</code> or editing <code>config.py</code> (paths/parameters/templates).</li>
    <li>After moving the project folder.</li>
  </ul>
  <b>Editing step code?</b> Re-import the step and <code>importlib.reload(module)</code> before calling <code>main()</code>.
</div>


In [1]:
# Bootstrap imports for package-style repo
import os, sys, importlib
from pathlib import Path

# >>>> EDIT THIS ONLY <<<<
AOI_VALUE = "benguela"  # e.g., "benguela", "huambo", "bie", "moxico", "moxicoleste"

# Auto-detect project root (works from notebooks/ or repo root)
_here = Path(".").resolve()
PROJECT_ROOT = _here.parent if _here.name == "notebooks" else _here

# Environment for the codebase
os.environ["PROJECT_ROOT"] = str(PROJECT_ROOT)
os.environ["AOI"] = AOI_VALUE

src_path = PROJECT_ROOT / "src"
if str(src_path) not in sys.path:
    sys.path.append(str(src_path))

# Make sure Python sees any file edits we made
importlib.invalidate_caches()

from config import PATHS, PARAMS, AOI, get_logger
print("AOI:", AOI)
print("ROOT:", PROJECT_ROOT)


def run_step(name: str, reload: bool = False):
    """Import and run a step module's main(). Set reload=True after editing the .py file."""
    importlib.invalidate_caches()
    m = importlib.import_module(name)
    if reload:
        importlib.reload(m)
    m.main()


AOI: benguela
ROOT: /mnt/c/Users/benny/OneDrive/Documents/Github/ago-lobitocorridor-analysis


## Step 00 — Data alignment & vector→raster

Align everything to a single 1-km template so every pixel stacks perfectly. This step converts key vectors (cropland, electrification, settlements) to rasters, harmonizes units/CRS, and prepares a 1-km flood screening layer alongside the native 30 m flood depth for engineering checks.

**Objective.** Produce AOI-aligned base rasters and vector-to-raster layers for consistent downstream math.

**Inputs.**
* `PARAMS.TARGET_GRID` (template), raw rasters (POP/NTL/VEG/drought), flood depth (30 m), and vectors (cropland, electrification, settlements).
  
**Outputs.**
* `{AOI}_pop_1km.tif`, `{AOI}_ntl_1km.tif`, `{AOI}_veg_1km.tif`, `{AOI}_drought_1km.tif`,
* `{AOI}_cropland_1km.tif`, `{AOI}_electric_1km.tif`, `{AOI}_settlement_1km.tif`,
* `{AOI}_flood_maxdepth_30m.tif` + `{AOI}_flood_maxdepth_1km.tif`,
* `{AOI}_rwi_meta_1km.tif` (if available in your data bundle).
  
**Assumptions & design.**
* Everything is reprojected/matched to the template; vector rasterization uses `all_touched=True`.
* Nodata is respected; logging warns on missing optional layers.

**Knobs.** 
* In `config.py`: source paths, `PARAMS.TARGET_GRID`.


In [2]:
run_step("step_00_align_and_rasterize")

12:11:23 | INFO | Loaded target grid | CRS=EPSG:4326 | size=254x311 | cell=0.0083x0.0083 (units of CRS)
12:11:23 | INFO | Aligning rasters to target grid...
12:11:24 | INFO | Wrote benguela_pop_1km.tif
12:11:24 | INFO | Wrote benguela_ntl_1km.tif
12:11:25 | INFO | Wrote benguela_veg_1km.tif
12:11:25 | INFO | Wrote benguela_drought_1km.tif
12:11:25 | INFO | Aligning Meta Relative Wealth Index (RWI) to 1-km grid...
12:11:25 | INFO | Wrote benguela_rwi_meta_1km.tif
12:11:25 | INFO | Rasterizing cropland presence (1=any in cell)...
12:12:25 | INFO | Wrote benguela_cropland_presence_1km.tif
12:12:25 | INFO | Rasterizing cropland fraction (0..1 per 1-km cell, supersample=10)...
12:13:28 | INFO | Wrote benguela_cropland_fraction_1km.tif | mean_fraction=0.120
12:13:31 | INFO | Rasterizing electricity masks using field 'FinalElecC' (1=grid, 99=unelectrified)...
12:13:35 | INFO | Wrote benguela_elec_grid_1km.tif, benguela_elec_unelectrified_1km.tif
12:13:37 | INFO | Rasterizing settlement type u

## Step 01 — Isochrones (accumulated travel time)

Turn road/friction inputs into a travel-time surface in minutes. This becomes the backbone for coverage masks (≤30/60/120 min) and access KPIs.

**Objective.** Compute the accumulated cost (minutes) raster.
                                                                                                               
**Inputs.**
* `PARAMS.TARGET_GRID`, friction/network inputs per `config.py`.
  
**Outputs.**
* `{AOI}_traveltime_min.tif` (accumulated cost surface).
  
**Assumptions.**
* Minutes as the canonical unit; obstacles/penalties embedded in friction.
  
**Knobs.**
* Friction parameters in `PARAMS` (if exposed), thresholds in `PARAMS.ISO_THRESH` (e.g., `(30, 60, 120)`).


In [3]:
run_step("step_01_isochrones")

12:13:45 | INFO | Isochrones from ago_phy_benguela_traveltime_market.tif | CRS=EPSG:4326 | size=254x311 | cell=0.0x0.0
12:13:45 | INFO | Building ≤30 min isochrone...
12:13:45 | INFO | Wrote benguela_iso_le_30min_1km.tif | cells inside: 2010/47177 (4.3%)
12:13:45 | INFO | Building ≤60 min isochrone...
12:13:45 | INFO | Wrote benguela_iso_le_60min_1km.tif | cells inside: 5902/47177 (12.5%)
12:13:45 | INFO | Building ≤120 min isochrone...
12:13:45 | INFO | Wrote benguela_iso_le_120min_1km.tif | cells inside: 17975/47177 (38.1%)
12:13:45 | INFO | Building ≤240 min isochrone...
12:13:45 | INFO | Wrote benguela_iso_le_240min_1km.tif | cells inside: 37758/47177 (80.0%)
12:13:45 | INFO | Step 01 complete.


## Step 02 — KPIs for isochrone coverage (population, cropland, electrification)

Quantify how much population, cropland, and electrification lie within each time threshold. These coverage KPIs are the simplest way to communicate “who’s connected.”

**Objective.** Summarize POP/CROP/ELECTRIC inside `tt ≤ threshold`.

**Inputs.**
* `{AOI}_traveltime_min.tif`, 1-km POP/CROPLAND/ELECTRIC rasters.
  
**Outputs.**
* `tables/{AOI}_iso_kpis.csv` (per threshold).
  
**Assumptions.**
* Coverage mask: `tt_minutes <= threshold`.
  
**Knobs.**
* `PARAMS.ISO_THRESH` list.


In [4]:
run_step("step_02_kpis_population_cropland_electric")

12:13:45 | INFO | Electrification grid value counts (non-NaN): {np.int64(1): 722}
12:13:45 | INFO | KPI base rasters loaded | CRS=EPSG:4326 | size=254x311 | cell=0.0083x0.0083
12:13:45 | INFO | Electrification mapping: FinalElecCode2020 -> elec_grid_1km == 1 (grid); 99/0/NaN = non-grid.
12:13:45 | INFO | Denominators | pop_total=3,271,805 | cropland_total_km2=5,749.06 | electrified_cells_total=722 | cell_area_km2≈1.000
12:13:45 | INFO | ≤ 30 min | cells=2,010 (area=2,010.00 km²) | pop=1,857,610 (56.8%) | crop=490.53 km² (8.5%) | elec_cells=541 (74.9%)
12:13:45 | INFO | ≤ 60 min | cells=5,902 (area=5,902.00 km²) | pop=2,157,382 (65.9%) | crop=1,225.03 km² (21.3%) | elec_cells=605 (83.8%)
12:13:45 | INFO | ≤120 min | cells=17,975 (area=17,975.00 km²) | pop=2,683,654 (82.0%) | crop=2,946.54 km² (51.3%) | elec_cells=662 (91.7%)
12:13:45 | INFO | ≤240 min | cells=37,758 (area=37,758.00 km²) | pop=3,175,567 (97.1%) | crop=5,272.59 km² (91.7%) | elec_cells=680 (94.2%)
12:13:45 | INFO | Saved 

## Step 03 — Priority surface (baseline)

Combine normalized indicators: better access (lower minutes), higher population & vegetation, stronger nighttime lights, and lower drought frequency. Save the score raster and a top-decile mask to highlight quick-win areas.

**Objective.** (Baseline) Build a transparent, minimal priority surface from core overlays.

**Inputs.**
* 1-km standardized layers (POP/CROP/ELECTRIC etc.).
  
**Outputs.**
* `rasters/{AOI}_priority_score_0_1.tif`, `rasters/{AOI}_priority_top10_mask.tif`.
  
**Assumptions.**
* Simpler normalization and weights vs Step 07.
  
**Recommendation.**
* Keep Step 03 as **baseline** only; prefer Step 07 for production.


In [5]:
run_step("step_03_priority_surface")

12:13:46 | INFO | Priority surface inputs loaded | CRS=EPSG:4326 | size=254x311 | cell=0.0083x0.0083
12:13:46 | INFO | Weights | ACC=0.35 POP=0.25 VEG=0.20 NTL=0.10 DRT=0.10
12:13:46 | INFO | Wrote benguela_priority_score_v1_0_1.tif
12:13:46 | INFO | Wrote benguela_priority_top10_mask_v1.tif | P90=0.656 | top10 cells: 1257/12568 (10.0%)
12:13:46 | INFO | Step 03 complete.


## Step 04 — Flood bottlenecks from road raster

Flag where roads intersect flood depths beyond thresholds. Use the 30 m layer for engineering plausibility, and the 1 km “max depth” for corridor-scale screening.

**Objective.** Identify likely flood bottlenecks impacting road access.

**Inputs.**
* 30 m flood depth, road raster or vector.
  
**Outputs.**
* `rasters/{AOI}_flood_bottlenecks_1km.tif` (or CSV summary, depending on implementation).
  
**Assumptions.**
* Engineering check: depth thresholds configurable.
  
**Knobs.**
* Flood depth threshold(s) in `PARAMS` (if exposed).


In [6]:
run_step("step_04_flood_bottlenecks_from_road_raster")

12:13:46 | WARNING | Reprojected to match target grid: flood1k, flood_frac, prio10
12:13:46 | INFO | Flood screening inputs loaded | CRS=EPSG:4326 | size=254x311 | cell=0.0083x0.0083
12:13:46 | INFO | Params | FLOOD_DEPTH_RISK=0.3 m | ROAD_CLASSES_KEEP=None | ROADS_ALL_TOUCHED=False
12:13:46 | INFO | Rasterizing roads: no class filter (use all OSM fclass values).
12:13:48 | INFO | Wrote benguela_roads_main_1km.tif | road_cells=11,886
12:13:48 | INFO | Wrote benguela_roads_flood_risk_cells_1km.tif | risk_cells=2,414
12:13:48 | INFO | Wrote benguela_roads_flood_risk_near_priority_1km.tif | risk_near_priority_cells=706
12:13:48 | INFO | Saved summary → /mnt/c/Users/benny/OneDrive/Documents/Github/ago-lobitocorridor-analysis/outputs/tables/benguela_roads_flood_risk_summary.csv
12:13:48 | INFO | Step 04 complete.


## Step 05 — Site audit points

Build “site cards” by sampling rasters around project points. Quicklook stats and thumbnails help triage which locations deserve deeper design work.

**Objective.** Generate site-level diagnostic metrics and snapshots.
    
**Inputs.**
* Project points (CSV/GeoPackage as per `config.py`).
  
**Outputs.**
* `tables/{AOI}_site_audit_points.csv` (+ optional PNGs in `outputs/figs/`).
  
**Assumptions.**
* Points in EPSG:4326; rasters sampled with nearest or bilinear.
  
**Knobs.**
* Sampling radius / list of rasters to sample (if exposed).


In [7]:
run_step("step_05_site_audit_points")

12:13:49 | INFO | Site audit inputs loaded | CRS=EPSG:4326 | size=254x311 | cell=0.0083x0.0083
12:13:49 | INFO | Loaded 26 site(s) from ago_poi_benguela_projectloc_dm_p.shp
12:13:49 | INFO | Neighborhood kernel: radius=5 cell(s) | cells_in_kernel=81
12:13:49 | INFO | Saved site audit → /mnt/c/Users/benny/OneDrive/Documents/Github/ago-lobitocorridor-analysis/outputs/tables/benguela_site_audit_points.csv | rows=25 | skipped_outside_grid=1
12:13:49 | INFO | Step 05 complete.


/home/bennyistanto/miniforge3/envs/cli/lib/python3.12/site-packages/rasterio/transform.py:529: RuntimeWarning: overflow encountered in dot
  transform.dot(input_matrix, out=input_matrix)
/home/bennyistanto/miniforge3/envs/cli/lib/python3.12/site-packages/rasterio/transform.py:410: RuntimeWarning: invalid value encountered in cast
  new_rows = np.floor(new_rows).astype(dtype="int32")
/home/bennyistanto/miniforge3/envs/cli/lib/python3.12/site-packages/rasterio/transform.py:411: RuntimeWarning: invalid value encountered in cast
  new_cols = np.floor(new_cols).astype(dtype="int32")


## Step 06 — Municipality (Admin2) ingest & correlations

Ingest Admin2 (RAPP) themes, standardize units, rasterize selected variables, and test correlations versus rural poverty. Produces a tidy “profiles” table and 1-km Admin2 surfaces usable in later steps.

**Objective.** Normalize Admin2 indicators and rasterize key variables.
    
**Inputs.**
* `data/vectors/ago_*_rapp_2020_a.shp` (themes in `THEME_VARS`), `PARAMS.TARGET_GRID`.
  
**Outputs.**
* `tables/{AOI}_municipality_indicators.csv` (wide),
* `tables/{AOI}_corr_with_rural_poverty.csv`,
* `tables/{AOI}_municipality_profiles.csv`,
* `rasters/{AOI}_muni_{theme}_{var}_1km.tif`.
  
**Assumptions & design.**
* Percentages normalized to 0–1 (`RAPP_PCT_IS_0_100=True`); travel time hours → minutes; `all_touched=True`.
* Missing variables skipped with warnings.
  
**Knobs.**
* `THEME_VARS` mapping, list of variables to rasterize.


In [8]:
run_step("step_06_muni_ingest")

12:13:49 | INFO | Target grid loaded | CRS=EPSG:4326 | size=254x311 | cell=0.0083x0.0083
12:13:49 | INFO | Step 06 will skip themes: ('climevents',)
12:13:49 | INFO | Theme=waterresources: read=0.06s, rename=0.00s, normalize=0.00s, rows=10, cols=10
12:13:49 | INFO | Theme=communications: read=0.07s, rename=0.00s, normalize=0.01s, rows=10, cols=11
12:13:49 | INFO | Theme=infra: read=0.08s, rename=0.00s, normalize=0.01s, rows=10, cols=15
12:13:49 | INFO | Theme=foodinsecurity: read=0.08s, rename=0.00s, normalize=0.01s, rows=10, cols=14
12:13:49 | INFO | Theme=outflow: read=0.08s, rename=0.00s, normalize=0.00s, rows=10, cols=12
12:13:49 | INFO | Theme=poverty: read=0.08s, rename=0.00s, normalize=0.00s, rows=10, cols=8
12:13:49 | INFO | Theme=productions: read=0.07s, rename=0.00s, normalize=0.01s, rows=10, cols=15
12:13:50 | INFO | Traveltime theme: converted 'avg_hours_to_market_financial' (hours) → 'avg_minutes_to_market_financial' (minutes).
12:13:50 | INFO | Theme=traveltime: read=0.06

## Step 07 — Priority surface (tunable, authoritative)

Create the **production** priority surface with coherent normalization, optional equity/poverty overlays, focal smoothing, tiny-cluster removal, and a Top-X% mask.

**Objective.** Build the authoritative priority map and mask with flexible weights.

**Inputs.**
* 1-km base layers + optional Admin2 rasters (auto-discovered), including `rwi_meta_1km` if present.
  
**Outputs.**
* `rasters/{AOI}_priority_score_0_1.tif`, `rasters/{AOI}_priority_top10_mask.tif`.
  
**Assumptions & design.**
* Unified normalization, masks, focal smoothing, small-cluster removal, Top-X selection.
  
**Knobs.**
* Weights and Top-X% in `PARAMS`/module (e.g., `PARAMS.TOP_X_PCT`).
  
**Note.**
* Downstream steps (11, 12) expect **these** canonical filenames from `config.py`.


In [9]:
run_step("step_07_priority_tunable")

12:13:56 | INFO | Target grid | CRS=EPSG:4326 | size=254x311 | cell=0.0083x0.0083
12:13:59 | INFO | Priority weight blend → ACC:0.29, POP:0.21, DRT:0.08, POV:0.12, FOOD:0.08, MTT:0.08, RWI:0.12
12:13:59 | INFO | Wrote benguela_priority_score_0_1.tif
12:13:59 | INFO | Wrote benguela_priority_top10_mask.tif | selected=515 cells
12:13:59 | INFO | Wrote sidecar meta → benguela_priority_score_0_1.meta.json
12:14:00 | INFO | Wrote Admin-2 priority table → benguela_priority_admin2_rank.csv (rows=10, cols=13)
12:14:00 | INFO | Step 07 complete.


## Step 08 — Project-level KPIs

Aggregate access/exposure/equity metrics over project buffers or polygons. Useful for comparing alternatives under the same data stack.

**Objective.** Compute KPI summaries for each project footprint/buffer.
    
**Inputs.**
* Project geometries + aligned rasters (POP/CROP/ELECTRIC/PRIORITY/etc.).
  
**Outputs.**
* `tables/{AOI}_project_kpis.csv`.
  
**Assumptions.**
* Buffers in meters; stats computed with consistent nodata handling.
  
**Knobs.**
* Buffer distance; KPI list (in step module).


In [10]:
run_step("step_08_project_kpis")

12:14:00 | INFO | Template grid | CRS=EPSG:4326 | size=254x311 | cell=0.0083x0.0083
12:14:00 | INFO | Denominators | pop_total=3,271,805 | cropland_total_km2=4821.60 | electrified_cells_total=722 | cell_area_km2=0.839
12:14:00 | INFO | Loaded 26 project(s) from ago_poi_benguela_projectloc_dm_p.shp
12:14:00 | INFO | Loaded 26 project(s) from ago_poi_benguela_projects_wb_p.shp
12:14:01 | INFO | Saved project KPIs → benguela_project_kpis.csv | rows=52
12:14:01 | INFO | Step 08 complete.


## Step 09 — Municipality (Admin2) targeting

Rank Admin2 units using access (≤60/120 min), poverty/equity signals, and priority shares. Produces an actionable shortlist for piloting.

**Objective.** Score and rank Admin2s for investment targeting.

**Inputs.**
* Admin2 rasters and tabular indicators from Step 06 + travel-time surface.
  
**Outputs.**
* `tables/{AOI}_muni_targeting.csv` (+ optional choropleth figs).
  
**Assumptions.**
* Uses `%≤60/≤120m` coverage, priority mask share, etc.
  
**Knobs.**
* Thresholds in `PARAMS.ISO_THRESH`; ranking weights in the module.


In [11]:
run_step("step_09_muni_targeting")

12:14:01 | INFO | Template grid | CRS=EPSG:4326 | size=254x311 | cell=0.0083x0.0083
12:14:01 | INFO | Using admin2 geometry from ago_pop_benguela_adm2_poverty_rapp_2020_a.shp
12:14:02 | INFO | Saved municipality targeting table → benguela_priority_muni_rank.csv | rows=10
12:14:02 | INFO | Wrote benefit incidence curve -> benguela_benefit_incidence.csv
12:14:02 | INFO | Equity index = 0.5961 (pro-poor) -> benguela_equity_summary.csv
12:14:02 | INFO | Step 09 complete.


## Step 10 — Priority scenarios

Stress-test the priority model with alternative weight sets (e.g., stronger equity vs. market access) and compare outcomes.

**Objective.** Generate and record scenario variants of the priority surface.
    
**Inputs.**
* Same base overlays as Step 07, different weight/toggle sets.
  
**Outputs.**
* `rasters/{AOI}_priority_score_scnX.tif`, masks, and a `tables/{AOI}_priority_scenarios.csv`.
  
**Assumptions.**
* Scenarios defined in code or `PARAMS`.
  
**Knobs.**
* Scenario configs (weights/toggles) in the module or `PARAMS`.


In [12]:
run_step("step_10_priority_scenarios")

12:14:02 | INFO | Target grid | CRS=EPSG:4326 | size=254x311 | cell=0.0083x0.0083
12:14:02 | INFO | Using on-disk baseline: benguela_priority_top10_mask.tif
12:14:02 | INFO | Scenario [baseline] — ACC+POP+DRT, rural-only, min crop 5%, 3x3 smooth, Top10%
12:14:03 | INFO | Scenario [no_drought] — ACC+POP only; rural-only; min crop 5%; Top10%
12:14:03 | INFO | Scenario [veg_signal] — ACC+POP+VEG; rural-only; min crop 10%; Top10%
12:14:03 | INFO | Scenario [top_800km2] — ACC+POP+DRT, rural-only; select fixed 800 km²; mild smooth
12:14:03 | INFO | Wrote scenario summary → benguela_priority_scenarios_summary.csv | scenarios=4 | masks_saved=4
12:14:03 | INFO | Wrote scenarios sidecar → benguela_priority_scenarios.meta.json
12:14:03 | INFO | Step 10 complete.


## Step 11 — Priority clusters

Convert the Top-X% priority mask into connected clusters and compute cluster KPIs (population, cropland, RWI, and share within ≤30/60/120 min). Small, noisy blobs are removed using cell and km² thresholds.

**Objective.** Extract cluster polygons/labels and summarize them for shortlisting.
    
**Inputs.**
* `PRIORITY_TOP10_TIF` from `config.py`, `traveltime_min` + base rasters, optional `rwi_meta_1km`.
  
**Outputs.**
* `rasters/{AOI}_priority_clusters_1km.tif` (via `PRIORITY_CLUSTERS_TIF`),
* `tables/{AOI}_priority_clusters.csv`.
  
**Assumptions & design.**
* Prune tiny clusters by min cells and **km²** (now configurable via `PARAMS.MIN_CLUSTER_KM2`).
* Added KPIs: `share_le_30m/60m/120m` (uses `PARAMS.ISO_THRESH`).
  
**Knobs.**
* `PARAMS.MIN_CLUSTER_CELLS`, `PARAMS.MIN_CLUSTER_KM2`, `PARAMS.ISO_THRESH`.


In [13]:
run_step("step_11_priority_clusters")

12:14:04 | INFO | Template grid | CRS=EPSG:4326 | size=254x311 | cell=0.0083x0.0083
12:14:04 | INFO | Connected components found: 4
12:14:04 | INFO | Pruned clusters: kept=4 | removed=0 by cells<30 or km2<2.0
12:14:04 | INFO | Saved clusters → benguela_priority_clusters.csv | clusters=4
12:14:04 | INFO | Step 11 complete.


## Step 12 — Travel-time catchments

Build catchment masks for each threshold and compute POP/CROP/RWI KPIs plus mean travel minutes—a quick validation of how “tight” the catchments are.

**Objective.** Derive per-threshold catchment KPIs and diagnostics.
    
**Inputs.**
* `{AOI}_traveltime_min.tif`, 1-km POP/CROPLAND, optional `rwi_meta_1km`.
  
**Outputs.**
* `tables/{AOI}_catchments_kpis.csv` (via `CATCHMENTS_KPI_CSV`),
* per-threshold masks in `rasters/` (if the step writes them).
  
**Assumptions & design.**
* Coverage = `tt <= thr`; includes `mean_travel_min` diagnostic.
  
**Knobs.**
* `PARAMS.ISO_THRESH`.


In [14]:
run_step("step_12_traveltime_catchments")

12:14:04 | INFO | Template grid | CRS=EPSG:4326 | size=254x311 | cell=0.0083x0.0083
12:14:04 | INFO | Building friction raster (minutes per km) from OSM roads + off-road baseline...
12:14:07 | INFO | Wrote benguela_friction_min_per_km.tif
12:14:07 | INFO | Computed isochrones for site 1 at 13.50298,-12.43023
12:14:07 | INFO | Computed isochrones for site 2 at 13.92658,-11.99189
12:14:07 | INFO | Computed isochrones for site 3 at 13.92658,-11.98361
12:14:08 | INFO | Computed isochrones for site 4 at 13.93564,-11.97778
12:14:08 | INFO | Computed isochrones for site 5 at 13.96986,-11.97678
12:14:08 | INFO | Computed isochrones for site 6 at 13.96986,-11.97169
12:14:08 | INFO | Computed isochrones for site 7 at 13.97575,-11.94578
12:14:09 | INFO | Computed isochrones for site 8 at 13.95425,-12.04464
12:14:09 | INFO | Computed isochrones for site 9 at 13.95228,-12.04692
12:14:09 | INFO | Computed isochrones for site 10 at 13.94531,-12.08803
12:14:10 | INFO | Computed isochrones for site 11 

/mnt/c/Users/benny/OneDrive/Documents/Github/ago-lobitocorridor-analysis/src/step_12_traveltime_catchments.py:277: RuntimeWarning: Mean of empty slice
  out["rwi_mean"] = float(np.nanmean(rwi_masked))


12:14:11 | INFO | Computed isochrones for site 14 at 13.85897,-11.81067
12:14:11 | INFO | Computed isochrones for site 15 at 13.91003,-12.15422
12:14:12 | INFO | Computed isochrones for site 16 at 14.06217,-11.86806
12:14:12 | INFO | Computed isochrones for site 17 at 14.02297,-12.11764
12:14:12 | INFO | Computed isochrones for site 18 at 14.63861,-12.08000
12:14:13 | INFO | Computed isochrones for site 19 at 13.84770,-12.30325
12:14:13 | INFO | Computed isochrones for site 20 at 14.80000,-12.91670
12:14:13 | INFO | Computed isochrones for site 21 at 14.76824,-12.35452
12:14:14 | INFO | Computed isochrones for site 22 at 14.00277,-13.02198


/home/bennyistanto/miniforge3/envs/cli/lib/python3.12/site-packages/rasterio/transform.py:529: RuntimeWarning: overflow encountered in dot
  transform.dot(input_matrix, out=input_matrix)
/home/bennyistanto/miniforge3/envs/cli/lib/python3.12/site-packages/rasterio/transform.py:410: RuntimeWarning: invalid value encountered in cast
  new_rows = np.floor(new_rows).astype(dtype="int32")
/home/bennyistanto/miniforge3/envs/cli/lib/python3.12/site-packages/rasterio/transform.py:411: RuntimeWarning: invalid value encountered in cast
  new_cols = np.floor(new_cols).astype(dtype="int32")
/mnt/c/Users/benny/OneDrive/Documents/Github/ago-lobitocorridor-analysis/src/step_12_traveltime_catchments.py:358: RuntimeWarning: Mean of empty slice
  row["mean_travel_min"] = float(np.nanmean(acc.values[msk.values > 0]))


12:14:14 | INFO | Computed isochrones for site 23 at -179769313486231570814527423731704356798070567525844996598917476803157260780028538760589558632766878171540458953514382464234321326889464182768467546703537516986049910576551282076245490090389328944075868508455133942304583236903222948165808559332123348274797826204144723168738177180919299881250404026184124858368.00000,-179769313486231570814527423731704356798070567525844996598917476803157260780028538760589558632766878171540458953514382464234321326889464182768467546703537516986049910576551282076245490090389328944075868508455133942304583236903222948165808559332123348274797826204144723168738177180919299881250404026184124858368.00000
12:14:14 | INFO | Computed isochrones for site 24 at 13.48618,-12.60254
12:14:15 | INFO | Computed isochrones for site 25 at 14.25963,-13.04825
12:14:15 | INFO | Computed isochrones for site 26 at 14.25974,-13.04839
12:14:15 | INFO | Catchments KPI: area monotone by site = 100%
12:14:15 | INFO | Wrote catchment 

/home/bennyistanto/miniforge3/envs/cli/lib/python3.12/site-packages/rasterio/transform.py:529: RuntimeWarning: overflow encountered in dot
  transform.dot(input_matrix, out=input_matrix)
/home/bennyistanto/miniforge3/envs/cli/lib/python3.12/site-packages/rasterio/transform.py:410: RuntimeWarning: invalid value encountered in cast
  new_rows = np.floor(new_rows).astype(dtype="int32")
/home/bennyistanto/miniforge3/envs/cli/lib/python3.12/site-packages/rasterio/transform.py:411: RuntimeWarning: invalid value encountered in cast
  new_cols = np.floor(new_cols).astype(dtype="int32")


## Step 13 — Synergies overlay

Intersect priority with constraints/opportunities (e.g., flood risk, grid proximity, conservation). Useful to surface “no-regret” or “high-risk” zones.

**Objective.** Produce an overlay table (and optional rasters) capturing key synergies/tradeoffs.

**Inputs.**
* Priority rasters, flood/constraint rasters, grid/market proximity if available.
  
**Outputs.**
* `tables/{AOI}_synergies_overlay.csv`, optionally composited rasters.
  
**Assumptions.**
* Binary or categorical overlay logic documented in the step.
  
**Knobs.**
* Overlay weights/toggles in the module.


In [15]:
run_step("step_13_synergies_overlay")

12:14:16 | INFO | Loaded 26 site(s).
12:14:16 | INFO | gov layer not found: ago_poi_benguela_projects_gov_p.shp
12:14:16 | INFO | Loaded 26 wb feature(s) from ago_poi_benguela_projects_wb_p.shp
12:14:16 | INFO | oth layer not found: ago_poi_benguela_projects_others_p.shp
12:14:16 | INFO | Saved site synergies → benguela_site_synergies.csv | rows=26
12:14:17 | INFO | Saved cluster synergies → benguela_cluster_synergies.csv | rows=4
12:14:17 | INFO | Step 13 complete.


/home/bennyistanto/miniforge3/envs/cli/lib/python3.12/site-packages/pandas/core/internals/blocks.py:1986: RuntimeWarning: overflow encountered in multiply
  values = self.values.round(decimals)  # type: ignore[union-attr]
/home/bennyistanto/miniforge3/envs/cli/lib/python3.12/site-packages/pandas/core/internals/blocks.py:1986: RuntimeWarning: overflow encountered in multiply
  values = self.values.round(decimals)  # type: ignore[union-attr]


## Step 14 — Origin-Destination (Admin2 gravity)

Sketch flows between Admin2 zones using a gravity model for corridor storytelling. Supports equity tilting (RWI), selectable impedance (exponential or power), distance cutoffs, and (optionally) doubly-constrained balancing.

**Objective.** Generate OD flows, zone attributes, and sampled agents for quick viz.
                                                                      
**Inputs.**
* `{AOI}_pop_1km.tif`, optional `rwi_meta_1km.tif`, Admin2 shapefile (`muni_path_for`).
  
**Outputs.**
* `tables/{AOI}_od_gravity.csv` (long: oi,dj,flow,dist_km),
* `tables/{AOI}_od_zone_attrs.csv` (zone masses + centroids),
* `tables/{AOI}_od_agents.csv` (sampled OD pairs for quick viz).
  
**Assumptions & design.**
* Geodesic centroid–centroid distances; **optional** RWI mass tilt;
* Impedance `f(D)` selectable (`exp` or `pow`); optional **doubly-constrained** via IPF; distance cutoff.
  
**Knobs (in `PARAMS`).**
* `OD_F`: `"exp"` or `"pow"`; `OD_LAMBDA` (exp), `OD_BETA` (pow),
* `OD_MAX_DIST_KM`, `OD_TRIPS_TOTAL`, `OD_USE_DOUBLY_CONSTRAINED`,
* `USE_RWI_IN_MASS`, `RWI_WEIGHT`, `OD_N_AGENTS`.


In [16]:
run_step("step_14_lite_od")

12:14:17 | INFO | RWI found: using equity tilt.
12:14:18 | INFO | OD zones aligned by ADM2CD_c order | N=10
12:14:18 | INFO | Flows: total=1,000,000 | mean_dist_km=75.3 | within_cutoff%=100.0
12:14:18 | INFO | Wrote benguela_od_gravity.csv, benguela_od_zone_attrs.csv
12:14:18 | INFO | Wrote benguela_od_agents.csv (N=100)
12:14:18 | INFO | OD-bottleneck: processing 51 OD pairs (90% of flow)...
12:14:18 | INFO | Wrote OD-bottleneck table -> benguela_od_bottleneck_cells.csv (106 risk cells with flow)
12:14:18 | INFO | Wrote benguela_od_bottleneck_risk.tif
12:14:18 | INFO | Step 14 complete.


## Step 15 — Corridor dashboard (cross-province summary)

Aggregates headline metrics from all processed AOIs into a single corridor-wide summary table and cluster inventory. This step loops over completed provinces—run it **after** all AOIs have finished Steps 00–14.

**Outputs:**
- `outputs/tables/corridor_dashboard.csv` — one row per AOI with top clusters, catchment reach, equity index, OD hubs
- `outputs/tables/corridor_cluster_inventory.csv` — all clusters across all provinces stacked

> **Note:** This step reads from *all* AOI outputs, not just the current one. It should be run once after all provinces are processed.

In [17]:
run_step("step_15_corridor_dashboard")

12:14:18 | INFO | Aggregating dashboard metrics for benguela ...
12:14:18 | INFO | Aggregating dashboard metrics for bie ...
12:14:18 | INFO | Aggregating dashboard metrics for huambo ...
12:14:19 | INFO | Aggregating dashboard metrics for moxico ...
12:14:19 | INFO | Aggregating dashboard metrics for moxicoleste ...
12:14:19 | INFO | Wrote corridor_dashboard.csv (6 rows)
12:14:19 | INFO | Wrote corridor_cluster_inventory.csv (7 clusters across all AOIs)
12:14:19 | INFO | Step 15 complete.


## Step 16 — Intervention simulator (road upgrade impact)

Simulates the impact of upgrading flood-vulnerable road segments to secondary-road standard (45 km/h). Recomputes travel times from all project sites and quantifies population gaining at each improvement threshold.

**Outputs:**
- `outputs/rasters/{AOI}_sim_travel_after.tif` — post-intervention travel surface
- `outputs/rasters/{AOI}_sim_travel_delta.tif` — minutes saved per cell (positive = improvement)
- `outputs/tables/{AOI}_sim_impact_summary.csv` — population gaining ≥5/10/15/30/60 min, plus newly within 60/120 min

> **Customisation:** To change the upgrade speed, edit `SIM_TARGET_SPEED_KMH` in `config.py`. To upgrade different road cells (not just flood-risk), supply a custom binary mask.

In [18]:
run_step("step_16_intervention_simulator")

12:14:19 | INFO | Sites: 25 project locations as cost-distance sources
12:14:19 | INFO | Computing baseline multi-source travel surface...
12:14:19 | INFO | Intervention: upgraded 2122 flood-risk road cells to 45 km/h
12:14:19 | INFO | Computing post-intervention travel surface...
12:14:20 | INFO | Wrote sim_travel_after.tif and sim_travel_delta.tif


/home/bennyistanto/miniforge3/envs/cli/lib/python3.12/site-packages/rasterio/transform.py:529: RuntimeWarning: overflow encountered in dot
  transform.dot(input_matrix, out=input_matrix)
/home/bennyistanto/miniforge3/envs/cli/lib/python3.12/site-packages/rasterio/transform.py:410: RuntimeWarning: invalid value encountered in cast
  new_rows = np.floor(new_rows).astype(dtype="int32")
/home/bennyistanto/miniforge3/envs/cli/lib/python3.12/site-packages/rasterio/transform.py:411: RuntimeWarning: invalid value encountered in cast
  new_cols = np.floor(new_cols).astype(dtype="int32")


12:14:20 | INFO | Wrote benguela_sim_impact_summary.csv | scenario=upgrade_risk_roads_to_45kmh
12:14:20 | INFO | Impact: 2835589 people gain travel time | avg saving = 6.8 min | max saving = 93.7 min
12:14:20 | INFO | Step 16 complete.


## End of Calculation

All 17 steps (00–16) complete. Outputs are in `outputs/rasters/` and `outputs/tables/`.

**Next:** Open `ago_lc_summary.ipynb` to generate summary tables S1–S7 and figures S1–S7 for the documentation.